common/scanner.h

```diff
@@ -1,294 +1,444 @@
-#include <tree_sitter/parser.h>
-#include <string>
+#include "tree_sitter/parser.h"
+#include <assert.h>
+#include <string.h>
 #include <wctype.h>
 
-namespace {
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
 
-enum {
-  COMMENT,
-  LEFT_QUOTED_STRING_DELIM,
-  RIGHT_QUOTED_STRING_DELIM,
-  STRING_DELIM,
-  LINE_NUMBER_DIRECTIVE,
-  NULL_CHARACTER
-};
+#define STRING_RESIZE(vec, _cap)                                               \
+    void *tmp = realloc((vec).data, ((_cap) + 1) * sizeof((vec).data[0]));     \
```

Comment:
Null terminator

Reply:
But we're not using null-terminated strings here. You know the length of a string becaused it's stored in `len`.

Reply:
Yes, but C library functions want that - in the context of this scanner, sure, we don't. But say someone uses strlen or strcpy in the future for the grammar, this would cause issues then

Reply:
Using `strlen()` here would be wrong. If you need the length of `quoted_string_id`, you should use the value stored in `len`. You're adding a lot of complexity to support something that I'd consider to be a bug.

Reply:
yeah, true, I guess strcmp/strcpy are more valid reasons but I'll remove that anyways